src/core/geometry/qgsbox3d.cpp

```diff
@@ -86,47 +112,110 @@ QgsBox3d QgsBox3d::intersect( const QgsBox3d &other ) const
 
 bool QgsBox3d::is2d() const
 {
-  return qgsDoubleNear( mZmin, mZmax ) || ( mZmin > mZmax );
+  return qgsDoubleNear( mZmin, mZmax ) || ( mZmin > mZmax ) || std::isnan( mZmin ) || std::isnan( mZmax ) ;
+}
+
+bool QgsBox3d::is3D() const
+{
+  return !is2d() && !isNull();
 }
 
 bool QgsBox3d::intersects( const QgsBox3d &other ) const
 {
   if ( !mBounds2d.intersects( other.mBounds2d ) )
     return false;
 
-  const double z1 = ( mZmin > other.mZmin ? mZmin : other.mZmin );
-  const double z2 = ( mZmax < other.mZmax ? mZmax : other.mZmax );
-  return z1 <= z2;
+  if ( other.is2d() || is2d() )
+  {
+    return true;
+  }
+  else
+  {
+    const double z1 = ( mZmin > other.mZmin ? mZmin : other.mZmin );
+    const double z2 = ( mZmax < other.mZmax ? mZmax : other.mZmax );
+    return z1 <= z2;
+  }
 }
 
 bool QgsBox3d::contains( const QgsBox3d &other ) const
 {
   if ( !mBounds2d.contains( other.mBounds2d ) )
     return false;
 
-  return ( other.mZmin >= mZmin && other.mZmax <= mZmax );
+  if ( other.is2d() || is2d() )
+  {
+    return true;
+  }
+  else
+  {
+    return ( other.mZmin >= mZmin && other.mZmax <= mZmax );
+  }
 }
 
 bool QgsBox3d::contains( const QgsPoint &p ) const
 {
-  if ( !mBounds2d.contains( p.x(), p.y() ) )
+  if ( p.is3D() && is3D() )
```

Comment:
I think it's safe to remove the `p.is3D()` here since you test if it's a nan in `contains(double, double, double)`.